src/dict.js

```diff
@@ -69,68 +99,66 @@ Sk.builtin.dict.tp$call = function(args, kw) {
 Sk.abstr.setUpInheritance("dict", Sk.builtin.dict, Sk.builtin.object);
 Sk.abstr.markUnhashable(Sk.builtin.dict);
 
-var kf = Sk.builtin.hash;
+var reg = /^[0-9!#_]/; // avoid clashes with complex key hashes 
+// and adjust for js ints not preserving order: str('1') => "!1"
 
-Sk.builtin.dict.prototype.key$lookup = function (bucket, key) {
-    var item;
-    var eq;
-    var i;
-
-    // Fast path: We spend a *lot* of time looking up strings
-    // in dictionaries. (Every attribute access, for starters.)
-    if (key.ob$type === Sk.builtin.str) {
-        for (i = 0; i < bucket.items.length; i++) {
-            item = bucket.items[i];
-            if (item.lhs.ob$type === Sk.builtin.str && item.lhs.v === key.v) {
-                return item;
-            }
-        }
-        return null;
-    }
-
-    for (i = 0; i < bucket.items.length; i++) {
-        item = bucket.items[i];
-        eq = Sk.misceval.richCompareBool(item.lhs, key, "Eq");
-        if (eq) {
-            return item;
-        }
+function kf(key) {
+    // str => jsstr().replace(/^[0-9!#_]/, "!$&") avoids conflicts
+    // other => hash.v value from builtin.hash (javascript number)
+    let key_hash = key.$savedKeyHash_; 
+    if (key_hash !== undefined) {
+        return key_hash;
+    } else if (key.ob$type === Sk.builtin.str) {
+        key_hash = key.$jsstr().replace(reg, "!$&"); 
+        key.$savedKeyHash_ = key_hash;
+        return key_hash;
+    } else if (typeof key === "string") {
+        // temporary while sysModules allows javascript strings as keys to python dicts
+        return key.replace(reg, "!$&");
     }
-
-    return null;
+    key_hash = Sk.builtin.hash(key).v; // builtin.hash returns an int
+    key.$savedKeyHash_ = key_hash;
+    return key_hash;
 };
 
-Sk.builtin.dict.prototype.key$pop = function (bucket, key) {
-    var item;
-    var eq;
-    var i;
+Sk.builtin.dict.prototype.sk$asarray = function () {
+    return Object.values(this.entries).map((x) => x.lhs);
+};
```

Comment:
Am I understanding correctly that this will not necessarily return all of the keys?  Just the ones that are stored in entries, and not in buckets?  If true, doesn't that break the ```sk$asarray``` "contract"?  (Resulting in incorrect behavior when other functions use ```sk$asarray``` which assume that it will correctly convert a dictionary to an array of all keys.)  Or am I missing something?

Reply:
`this.entries` has all the keys. So it abides by the contract. (I think if it didn't some tests should have broken [i hope])

from the example in the description
```python
d = {1:0, 1.1: 1, 1.2: 2, 'a' : 3, 'b': 4, 2: 5}
```
```javascript
  this.entries: 
    #1_0: {lhs: int_(1),     rhs: int_(0)} // non string (int_(1)
    #1_1: {lhs: float_(1.1), rhs: int_(1)} // non string
    #1_2: {lhs: float_(1.2), rhs: int_(2)} // non string
     a:   {lhs: str('a'),    rhs: int_(3)} //string 
     b:   {lhs: str('b'),    rhs: int_(4)} // string 
    #2_0: {lhs: int_(2),     rhs: int_(5)} 
```

`entries` contains all the items
`buckets` deals with clashes - in other words - everything except strings

Reply:
Rereading the comments, they do imply this, but the descriptions are a little unclear to me.  I feel like it would be helpful to be more explicit that all dictionary key/value pairs are in ```this.entries``` and they are **duplicated** in ```this.buckets``` if they are not string keys.  It would also be helpful, in my opinion, if the example dictionaries in the comments were the same dictionary in both cases.  In the ```entries``` description, you show the Python dictionary and the corresponding Skulpt representation.  In the ```buckets``` description, you use a different Python dictionary, don't show the Python dictionary, and then show the Skulpt representation of a single bucket without showing what was in entries.  If it were clearly the same dictionary, it would have been more obvious what was in entries, as that was already in the other comment.

Rereading it all, like I said, it does make sense.  I just think it could be much clearer.  Of course, I also could have just read more of the code, but I feel like the comments should speak for themselves....